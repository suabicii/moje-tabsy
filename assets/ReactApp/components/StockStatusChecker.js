import React from "react";
import {useSelector} from "react-redux";

function StockStatusChecker() {
    const drugList = useSelector(state => state.drugs);

    const runningOutOfStock = [];
    const outOfStock = [];
    const percentsToWarnUserAboutRunningOutOfStock = 30;
    drugList.forEach(drug => {
        if (parseInt(drug.quantity) === 0) {
            outOfStock.push(drug);
        } else if (drug.quantity * 100 / drug.quantityMax <= percentsToWarnUserAboutRunningOutOfStock) {
            runningOutOfStock.push(drug);
        }
    });

    if (!runningOutOfStock.length && !outOfStock.length) {
        return (
            <>
                <h5 className="card-title text-uppercase text-bg-success text-center rounded"
                    data-testid="stock-ok">Ok 👍</h5>
                <p className="card-text">Zapasy Twoich leków i suplementów są wystarczające 🤩</p>
            </>
        );
    } else {
        return (
            <>
                <h5 className="card-title text-uppercase text-bg-danger text-center rounded"
                    data-testid="stock-warning">Uwaga! 😱</h5>
                <ul className="list-unstyled">
                    {
                        runningOutOfStock.length > 0 &&
                        <p className="card-text">Zapasy następujących leków/suplementów ulegają wyczerpaniu:</p>
                    }
                    {runningOutOfStock.length > 0 && runningOutOfStock.map(drug =>
                        <li key={drug.id} className="text-danger" data-testid="drug-end">
                            {drug.name} – <strong>{drug.quantity}/{drug.quantityMax}</strong>
                        </li>
                    )}
                    {
                        outOfStock.length > 0 &&
                        <p className="card-text">Zapasy następujących leków/suplementów uległy wyczerpaniu:</p>
                    }
                    {outOfStock.length > 0 && outOfStock.map(drug =>
                        <li key={drug.id} className="text-bg-danger" data-testid="drug-end">
                            {drug.name} – <strong>{drug.quantity}/{drug.quantityMax}</strong>
                        </li>
                    )}
                </ul>
            </>
        );
    }
}

export default StockStatusChecker;