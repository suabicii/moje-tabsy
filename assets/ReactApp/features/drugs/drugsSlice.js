import {createSelector, createSlice} from "@reduxjs/toolkit";
import {fetchData} from "../../utils/fetchData";

const drugsSlice = createSlice({
    name: 'drugs',
    initialState: [],
    reducers: {
        addDrug: (state, action) => {
            const {id, name, quantity, quantityMax, unit, dosing, dosingMoments} = action.payload;
            state.push({id, name, quantity, quantityMax, unit, dosing, dosingMoments});
        },
        removeDrug: (state, action) => {
            const id = action.payload;
            return state.filter(drug => drug.id !== id);
        },
        editDrug: (state, action) => {
            const {id, ...updates} = action.payload;
            const drugIndex = state.findIndex(drug => drug.id === id);
            if (drugIndex !== -1) {
                state[drugIndex] = {...state[drugIndex], ...updates};
            }
        },
        setDrugs: (state, action) => action.payload
    },
    sort: true
});

const {reducer, actions} = drugsSlice;
export const {addDrug, removeDrug, editDrug, setDrugs} = actions;

const defaultDrugsSelector = state => state.drugs;

export const sortedDrugsSelector = createSelector(
    defaultDrugsSelector,
    drugs => [...drugs].sort((a, b) => a.name.localeCompare(b.name))
);

export const fetchDrugs = () => async dispatch => {
    await fetchData('/api/drug-list').then(data => {
        dispatch(setDrugs(data));
    }).catch(err => {
        console.log(err);
    });
};

export default reducer;