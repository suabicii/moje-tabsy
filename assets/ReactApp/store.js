import {configureStore} from "@reduxjs/toolkit";
import drugsSlice from "./features/drugs/drugsSlice"
import userSlice from "./features/user/userSlice";
import darkModeSlice from "./features/darkMode/darkModeSlice";

export default configureStore({
    reducer: {
        drugs: drugsSlice,
        user: userSlice,
        darkMode: darkModeSlice
    }
});