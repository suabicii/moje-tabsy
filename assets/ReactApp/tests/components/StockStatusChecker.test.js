/**
 * @jest-environment jsdom
 * */
import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import StockStatusChecker from "../../components/StockStatusChecker";
import {act, render, screen} from "@testing-library/react";
import drugs from "./fixtures/drugs";
import {Provider} from "react-redux";
import store from "../../store";
import {fetchDrugs} from "../../features/drugs/drugsSlice";

afterAll(() => {
    global.fetch.mockClear();
    delete global.fetch;
});

const renderStockStatusChecker = () => {
    render(
        <Provider store={store}>
            <StockStatusChecker/>
        </Provider>
    );
};

it('should correctly render StockStatusChecker component', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(
        <Provider store={store}>
            <StockStatusChecker/>
        </Provider>
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('should inform user about running out of stock', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
        json: () => Promise.resolve(drugs)
    }));
    store.dispatch(fetchDrugs());

    await act(() => {
        renderStockStatusChecker();
    });

    expect(screen.getByTestId('stock-warning')).toBeVisible();
    expect(screen.getByTestId('drug-end').textContent).toContain('Witamina C'); // From fixtures/drug.js, quantity 20/80
});

it('should inform user that everything is OK if there is enough drugs in storage', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
        json: () => Promise.resolve([drugs[0]])
    }));
    store.dispatch(fetchDrugs());

    await act(() => {
        renderStockStatusChecker();
    });

    expect(screen.getByTestId('stock-ok')).toBeVisible();
});