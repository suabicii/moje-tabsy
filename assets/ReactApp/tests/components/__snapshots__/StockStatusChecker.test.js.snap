// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render StockStatusChecker component 1`] = `
<Context.Provider
  value={
    Object {
      "addDrug": [Function],
      "drugList": Array [],
      "editDrug": [Function],
      "removeDrug": [Function],
    }
  }
>
  <StockStatusChecker />
</Context.Provider>
`;
